tfhe/src/integer/server_key/radix_parallel/rotate.rs

```diff
@@ -0,0 +1,665 @@
+    ///
+    /// let msg = 128u8;
+    /// let n = 2;
+    ///
+    /// let ct = cks.encrypt(msg as u64);
+    ///
+    /// let ct_res = sks.unchecked_scalar_rotate_left_parallelized(&ct, n);
+    ///
+    /// // Decrypt:
+    /// let dec: u64 = cks.decrypt(&ct_res);
+    /// assert_eq!(msg.rotate_left(n as u32) as u64, dec);
+    /// ```
+    pub fn unchecked_scalar_rotate_left_parallelized<PBSOrder: PBSOrderMarker>(
+        &self,
+        ct: &RadixCiphertext<PBSOrder>,
+        n: u64,
+    ) -> RadixCiphertext<PBSOrder> {
+        let mut result = ct.clone();
+        self.unchecked_scalar_rotate_left_assign_parallelized(&mut result, n);
+        result
+    }
+
+    /// Computes homomorphically a rotation of bits.
+    ///
+    /// Shifts the bits to the left by a specified amount,
+    /// `n`, wrapping the truncated bits to the end of the resulting integer.
+    ///
+    /// The result is assigned to the input ciphertext
+    ///
+    /// # Requirements
+    ///
+    /// - The blocks parameter's carry space have at least one more bit than message space
+    /// - The input ciphertext carry buffer is emtpy / clean
+    ///
+    /// # Output
+    ///
+    /// - The output's carries will be clean
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2;
+    ///
+    /// // We have 4 * 2 = 8 bits of message
+    /// let size = 4;
+    /// let (cks, sks) = gen_keys_radix(&PARAM_MESSAGE_2_CARRY_2, size);
+    ///
+    /// let msg = 128u8;
+    /// let n = 2;
+    ///
+    /// let mut ct = cks.encrypt(msg as u64);
+    ///
+    /// sks.unchecked_scalar_rotate_left_assign_parallelized(&mut ct, n);
+    ///
+    /// // Decrypt:
+    /// let dec: u64 = cks.decrypt(&ct);
+    /// assert_eq!(msg.rotate_left(n as u32) as u64, dec);
+    /// ```
+    pub fn unchecked_scalar_rotate_left_assign_parallelized<PBSOrder: PBSOrderMarker>(
+        &self,
+        ct: &mut RadixCiphertext<PBSOrder>,
+        n: u64,
+    ) {
+        // The general idea, is that we know by how much we want to
+        // rotate since `n` is a clear value.
+        //
+        // So we can use that to implement rotating in two step
+        // 1) rotate blocks
+        // 2) shift within each block and 'propagate' to the next one
+        debug_assert!(ct.block_carries_are_empty());
+        debug_assert!(self.key.carry_modulus.0 >= self.key.message_modulus.0 / 2);
+
+        let num_bits_in_block = self.key.message_modulus.0.ilog2() as u64;
+        let total_num_bits = num_bits_in_block * ct.blocks.len() as u64;
+
+        let n = n % total_num_bits;
+
+        if n == 0 {
+            return;
+        }
+
+        let rotations = (n / num_bits_in_block) as usize;
+        let shift_within_block = n % num_bits_in_block;
+        let num_blocks = ct.blocks.len();
+
+        dbg!(rotations);
+
+        // rotate left as the blocks are from LSB to MSB
```

Comment:
rotate right, there may be the same typo in the previous shift code

Reply:
I checked, shifts did not have this typo